File: June-LeetCoding-Challenge-2020/Week1/226-invert-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 * Runtime: 68 ms, faster than 24.58% of JavaScript online submissions for Invert Binary Tree.
 * Memory Usage: 33.9 MB, less than 20.00% of JavaScript online submissions for Invert Binary Tree.
 * TC: O(N), where N is the number of elements in a tree
 * SC: O(logN), where N is the number of elements in a tree
 */
var invertTree = function(root) {
    if (!root) {
        return null;
    }
    
    swapLeftRight(root);
    
    invertTree(root.left);
    invertTree(root.right);
    
    return root;
};

function swapLeftRight (node) {
    const tmp = node.left;
    node.left = node.right;
    node.right = tmp;
}

File: June-LeetCoding-Challenge-2020/Week1/226-invert-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    # Sol 1. Using Stack
    def invertTree(self, root: TreeNode) -> TreeNode:
        if not root:
            return None
            
        stack=[root]
        while stack:
            node = stack.pop()
            if node:
                node.left, node.right = node.right, node.left
                stack =stack+[node.left, node.right]
                
        return root
    # TC: O(N)
    # SC: O(N)
    # Runtime: 28 ms, faster than 75.77%
    # Memory: 14.1 MB, less than 5.41% 

class Solution:
    # Sol 2. Recursive Function
    def invertTree(self, root):
        if root is None:
            return None
        root.left, root.right = self.invertTree(root.right), self.invertTree(root.left)
        return root

    # TC: O(N)
    # SC: O(N)